File: routes/analytics.py
"""
Analytics API Routes
Provides endpoints for analytics data retrieval and business intelligence.
"""

from datetime import datetime, timedelta
from typing import List, Optional, Dict, Any
from fastapi import APIRouter, HTTPException, Query, status
from pydantic import BaseModel, Field

from services.analytics_service import get_analytics_service
from utils.logger import get_logger

logger = get_logger(__name__)
router = APIRouter(prefix="/analytics", tags=["Analytics"])

# Response Models
class UsageSummaryResponse(BaseModel):
    """Usage summary response model"""
    period_days: int
    total_requests: int
    total_tokens: int
    avg_response_time: float
    success_rate: float
    rag_usage_rate: float
    estimated_cost_savings: float
    top_models: List[Dict[str, Any]]

class ModelStatistics(BaseModel):
    """Model statistics response model"""
    model_name: str
    total_requests: int
    total_tokens: int
    avg_response_time: float
    success_rate: float
    last_used: Optional[str]
    total_errors: int
    avg_prompt_length: float
    avg_response_length: float
    tokens_per_second: float
    efficiency_score: float

class DocumentInsight(BaseModel):
    """Document insight response model"""
    document_id: str
    filename: str
    index_name: str
    query_count: int
    total_retrievals: int
    avg_relevance_score: float
    last_accessed: Optional[str]
    upload_date: Optional[str]
    chunk_count: int
    avg_chunk_retrieval: float
    retrieval_rate: float
    queries_per_day: float

class UsageTrends(BaseModel):
    """Usage trends response model"""
    daily_requests: List[Dict[str, Any]]
    hourly_pattern: List[Dict[str, Any]]
    weekly_growth_rate: float
    peak_hour: int
    peak_hour_requests: int

class BusinessReport(BaseModel):
    """Comprehensive business report model"""
    report_period_days: int
    generated_at: str
    executive_summary: Dict[str, Any]
    cost_analysis: Dict[str, Any]
    usage_metrics: Dict[str, Any]
    model_performance: List[Dict[str, Any]]
    document_insights: List[Dict[str, Any]]
    trends: Dict[str, Any]
    business_insights: List[str]
    recommendations: List[str]

class SystemHealth(BaseModel):
    """System health metrics"""
    cpu_usage: float
    memory_usage: float
    disk_usage: float
    gpu_usage: float
    gpu_memory: float
    active_requests: int
    status: str

# API Endpoints

@router.get(
    "/summary",
    response_model=UsageSummaryResponse,
    summary="Get usage summary",
    description="Get comprehensive usage summary for the specified time period."
)
async def get_usage_summary(
    days: int = Query(7, ge=1, le=365, description="Number of days to analyze")
) -> UsageSummaryResponse:
    """
    Get usage summary for the last N days.
    
    Includes:
    - Total requests and tokens
    - Average response time and success rate
    - RAG usage statistics
    - Cost savings estimation
    - Top performing models
    """
    try:
        analytics = get_analytics_service()
        summary = await analytics.get_usage_summary(days)
        
        if not summary:
            # Return empty summary if no data
            return UsageSummaryResponse(
                period_days=days,
                total_requests=0,
                total_tokens=0,
                avg_response_time=0.0,
                success_rate=100.0,
                rag_usage_rate=0.0,
                estimated_cost_savings=0.0,
                top_models=[]
            )
        
        return UsageSummaryResponse(**summary)
        
    except Exception as e:
        logger.error(f"Failed to get usage summary: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to retrieve usage summary: {str(e)}"
        )

@router.get(
    "/models",
    response_model=List[ModelStatistics],
    summary="Get model statistics",
    description="Get comprehensive statistics for all models including performance metrics."
)
async def get_model_statistics() -> List[ModelStatistics]:
    """
    Get detailed statistics for all models.
    
    Includes:
    - Usage counts and token statistics
    - Performance metrics (response time, success rate)
    - Efficiency scores and tokens per second
    - Error counts and reliability metrics
    """
    try:
        analytics = get_analytics_service()
        stats = await analytics.get_model_statistics()
        
        return [ModelStatistics(**stat) for stat in stats]
        
    except Exception as e:
        logger.error(f"Failed to get model statistics: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to retrieve model statistics: {str(e)}"
        )

@router.get(
    "/documents",
    response_model=List[DocumentInsight],
    summary="Get document insights",
    description="Get usage insights for documents including popularity and engagement metrics."
)
async def get_document_insights(
    limit: int = Query(20, ge=1, le=100, description="Maximum number of documents to return")
) -> List[DocumentInsight]:
    """
    Get document usage insights.
    
    Includes:
    - Query frequency and retrieval statistics
    - Relevance scores and engagement metrics
    - Popularity trends and usage patterns
    """
    try:
        analytics = get_analytics_service()
        insights = await analytics.get_document_insights(limit)
        
        return [DocumentInsight(**insight) for insight in insights]
        
    except Exception as e:
        logger.error(f"Failed to get document insights: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to retrieve document insights: {str(e)}"
        )

@router.get(
    "/trends",
    response_model=UsageTrends,
    summary="Get usage trends",
    description="Get usage trends and patterns over time including growth analysis."
)
async def get_usage_trends(
    days: int = Query(30, ge=7, le=365, description="Number of days to analyze")
) -> UsageTrends:
    """
    Get usage trends and patterns.
    
    Includes:
    - Daily request patterns
    - Hourly usage distribution
    - Growth rate analysis
    - Peak usage identification
    """
    try:
        analytics = get_analytics_service()
        trends = await analytics.get_usage_trends(days)
        
        if not trends:
            return UsageTrends(
                daily_requests=[],
                hourly_pattern=[],
                weekly_growth_rate=0.0,
                peak_hour=12,
                peak_hour_requests=0
            )
        
        return UsageTrends(**trends)
        
    except Exception as e:
        logger.error(f"Failed to get usage trends: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to retrieve usage trends: {str(e)}"
        )

@router.get(
    "/business-report",
    response_model=BusinessReport,
    summary="Generate business report",
    description="Generate comprehensive business intelligence report with ROI analysis."
)
async def generate_business_report(
    days: int = Query(30, ge=1, le=365, description="Number of days to analyze")
) -> BusinessReport:
    """
    Generate comprehensive business intelligence report.
    
    Includes:
    - Executive summary with key metrics
    - Cost analysis and ROI calculations
    - Performance insights and recommendations
    - Growth trends and usage patterns
    """
    try:
        analytics = get_analytics_service()
        report = await analytics.generate_business_report(days)
        
        if not report:
            # Return minimal report if no data
            return BusinessReport(
                report_period_days=days,
                generated_at=datetime.now().isoformat(),
                executive_summary={},
                cost_analysis={},
                usage_metrics={},
                model_performance=[],
                document_insights=[],
                trends={},
                business_insights=[],
                recommendations=["Start using the platform to generate analytics data"]
            )
        
        return BusinessReport(**report)
        
    except Exception as e:
        logger.error(f"Failed to generate business report: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to generate business report: {str(e)}"
        )

@router.get(
    "/health",
    response_model=SystemHealth,
    summary="Get system health",
    description="Get current system health and performance metrics."
)
async def get_system_health() -> SystemHealth:
    """
    Get current system health metrics.
    
    Includes:
    - CPU, memory, and disk usage
    - GPU utilization (if available)
    - Active request count
    - Overall system status
    """
    try:
        analytics = get_analytics_service()
        
        # Get metrics directly without the expensive collect_system_metrics call
        import psutil
        
        # Use non-blocking CPU measurement
        cpu_percent = psutil.cpu_percent(interval=None)
        memory = psutil.virtual_memory()
        disk = psutil.disk_usage('/')
        
        # Determine system status
        status = "healthy"
        if cpu_percent > 80 or memory.percent > 85 or (disk.used / disk.total * 100) > 90:
            status = "warning"
        if cpu_percent > 95 or memory.percent > 95 or (disk.used / disk.total * 100) > 95:
            status = "critical"
        
        return SystemHealth(
            cpu_usage=round(cpu_percent, 2),
            memory_usage=round(memory.percent, 2),
            disk_usage=round((disk.used / disk.total * 100), 2),
            gpu_usage=0.0,  # Would need GPUtil for actual GPU metrics
            gpu_memory=0.0,
            active_requests=len(analytics._active_sessions),
            status=status
        )
        
    except Exception as e:
        logger.error(f"Failed to get system health: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to retrieve system health: {str(e)}"
        )

@router.get(
    "/export",
    summary="Export analytics data",
    description="Export analytics data in JSON or CSV format."
)
async def export_analytics_data(
    format: str = Query("json", regex="^(json|csv)$", description="Export format: json or csv"),
    days: int = Query(30, ge=1, le=365, description="Number of days to export")
):
    """
    Export analytics data in specified format.
    
    Supports:
    - JSON format: Complete structured data
    - CSV format: Simplified tabular data
    """
    try:
        analytics = get_analytics_service()
        data = await analytics.export_data(format, days)
        
        if data is None:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Unsupported export format: {format}"
            )
        
        # Set appropriate content type
        media_type = "application/json" if format == "json" else "text/csv"
        filename = f"analytics_{datetime.now().strftime('%Y%m%d')}_{days}days.{format}"
        
        from fastapi.responses import Response
        
        return Response(
            content=data,
            media_type=media_type,
            headers={"Content-Disposition": f"attachment; filename={filename}"}
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to export analytics data: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to export data: {str(e)}"
        )

@router.post(
    "/cleanup",
    summary="Cleanup old data",
    description="Remove analytics data older than specified days to manage database size."
)
async def cleanup_analytics_data(
    days_to_keep: int = Query(90, ge=7, le=365, description="Days of data to keep")
):
    """
    Clean up old analytics data.
    
    Removes:
    - Request logs older than specified days
    - System metrics older than specified days
    - Maintains model and document statistics
    """
    try:
        analytics = get_analytics_service()
        await analytics.cleanup_old_data(days_to_keep)
        
        return {
            "success": True,
            "message": f"Cleaned up analytics data older than {days_to_keep} days",
            "timestamp": datetime.now().isoformat()
        }
        
    except Exception as e:
        logger.error(f"Failed to cleanup analytics data: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to cleanup data: {str(e)}"
        )

@router.get(
    "/realtime",
    summary="Get real-time metrics",
    description="Get real-time system and usage metrics for dashboard updates."
)
async def get_realtime_metrics():
    """
    Get real-time metrics for live dashboard updates.
    
    Returns:
    - Current system performance
    - Active session count
    - Recent request statistics
    """
    try:
        analytics = get_analytics_service()
        
        # Get summary for last hour
        summary = await analytics.get_usage_summary(days=1)  # Last 24 hours
        
        # Get system metrics
        health = await get_system_health()
        
        # Recent activity (last hour)
        recent_requests = summary.get('total_requests', 0)
        
        return {
            "timestamp": datetime.now().isoformat(),
            "requests_last_24h": recent_requests,
            "tokens_last_24h": summary.get('total_tokens', 0),
            "avg_response_time": summary.get('avg_response_time', 0.0),
            "success_rate": summary.get('success_rate', 100.0),
            "active_sessions": len(analytics._active_sessions),
            "system_health": {
                "cpu_usage": health.cpu_usage,
                "memory_usage": health.memory_usage,
                "disk_usage": health.disk_usage,
                "status": health.status
            }
        }
        
    except Exception as e:
        logger.error(f"Failed to get realtime metrics: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to retrieve realtime metrics: {str(e)}"
        )

File: frontend/src/services/analytics.ts
import axios from 'axios';

// Use relative URLs so requests go through Vite proxy
const API_BASE_URL = '';

const apiClient = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
  },
});

export interface UsageSummary {
  period_days: number;
  total_requests: number;
  total_tokens: number;
  avg_response_time: number;
  success_rate: number;
  rag_usage_rate: number;
  estimated_cost_savings: number;
  top_models: Array<{
    model: string;
    count: number;
  }>;
}

export interface ModelStatistics {
  model_name: string;
  total_requests: number;
  total_tokens: number;
  avg_response_time: number;
  success_rate: number;
  last_used?: string;
  total_errors: number;
  avg_prompt_length: number;
  avg_response_length: number;
  tokens_per_second: number;
  efficiency_score: number;
}

export interface DocumentInsight {
  document_id: string;
  filename: string;
  index_name: string;
  query_count: number;
  total_retrievals: number;
  avg_relevance_score: number;
  last_accessed?: string;
  upload_date?: string;
  chunk_count: number;
  avg_chunk_retrieval: number;
  retrieval_rate: number;
  queries_per_day: number;
}

export interface UsageTrends {
  daily_requests: Array<{
    date: string;
    requests: number;
  }>;
  hourly_pattern: Array<{
    hour: number;
    requests: number;
  }>;
  hourly_response_times?: Array<{
    hour: number;
    avg_response_time: number;
    request_count: number;
  }>;
  weekly_growth_rate: number;
  peak_hour: number;
  peak_hour_requests: number;
}

export interface BusinessReport {
  report_period_days: number;
  generated_at: string;
  executive_summary: {
    total_requests: number;
    total_tokens: number;
    avg_tokens_per_request: number;
    success_rate: number;
    total_cost_savings: number;
    monthly_projected_savings: number;
    roi_percentage: number;
  };
  cost_analysis: {
    local_hosting_cost_monthly: number;
    cloud_cost_comparisons: Record<string, number>;
    total_potential_cloud_cost: number;
    savings_percentage: number;
  };
  usage_metrics: UsageSummary;
  model_performance: ModelStatistics[];
  document_insights: DocumentInsight[];
  trends: UsageTrends;
  business_insights: string[];
  recommendations: string[];
}

export interface SystemHealth {
  cpu_usage: number;
  memory_usage: number;
  disk_usage: number;
  gpu_usage: number;
  gpu_memory: number;
  active_requests: number;
  status: string;
}

export interface RealtimeMetrics {
  timestamp: string;
  requests_last_24h: number;
  tokens_last_24h: number;
  avg_response_time: number;
  success_rate: number;
  active_sessions: number;
  system_health: {
    cpu_usage: number;
    memory_usage: number;
    disk_usage: number;
    status: string;
  };
}

class AnalyticsService {
  async getUsageSummary(days: number = 7): Promise<UsageSummary> {
    const response = await apiClient.get(`/analytics/summary?days=${days}`);
    return response.data;
  }

  async getModelStatistics(): Promise<ModelStatistics[]> {
    const response = await apiClient.get('/analytics/models');
    return response.data;
  }

  async getDocumentInsights(limit: number = 20): Promise<DocumentInsight[]> {
    const response = await apiClient.get(`/analytics/documents?limit=${limit}`);
    return response.data;
  }

  async getUsageTrends(days: number = 30): Promise<UsageTrends> {
    const response = await apiClient.get(`/analytics/trends?days=${days}`);
    return response.data;
  }

  async getBusinessReport(days: number = 30): Promise<BusinessReport> {
    const response = await apiClient.get(`/analytics/business-report?days=${days}`);
    return response.data;
  }

  async getSystemHealth(): Promise<SystemHealth> {
    const response = await apiClient.get('/analytics/health');
    return response.data;
  }

  async getRealtimeMetrics(): Promise<RealtimeMetrics> {
    const response = await apiClient.get('/analytics/realtime');
    return response.data;
  }

  async exportData(format: 'json' | 'csv' = 'json', days: number = 30): Promise<Blob> {
    const response = await apiClient.get(`/analytics/export?format=${format}&days=${days}`, {
      responseType: 'blob'
    });
    return response.data;
  }

  async cleanupData(daysToKeep: number = 90): Promise<{ success: boolean; message: string }> {
    const response = await apiClient.post(`/analytics/cleanup?days_to_keep=${daysToKeep}`);
    return response.data;
  }

  // Helper methods for data processing
  processModelData(models: ModelStatistics[]): Array<{
    model: string;
    count: number;
    percentage: number;
  }> {
    const totalRequests = models.reduce((sum, model) => sum + model.total_requests, 0);
    
    return models.map(model => ({
      model: model.model_name,
      count: model.total_requests,
      percentage: totalRequests > 0 ? Math.round((model.total_requests / totalRequests) * 100) : 0
    }));
  }

  formatCurrency(amount: number): string {
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
      minimumFractionDigits: 2,
      maximumFractionDigits: 2
    }).format(amount);
  }

  formatNumber(num: number): string {
    return new Intl.NumberFormat('en-US').format(num);
  }

  formatBytes(bytes: number): string {
    if (bytes === 0) return '0 Bytes';
    const k = 1024;
    const sizes = ['Bytes', 'KB', 'MB', 'GB'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i];
  }

  formatDuration(seconds: number): string {
    if (seconds < 1) return `${Math.round(seconds * 1000)}ms`;
    if (seconds < 60) return `${seconds.toFixed(1)}s`;
    const minutes = Math.floor(seconds / 60);
    const remainingSeconds = Math.round(seconds % 60);
    return `${minutes}m ${remainingSeconds}s`;
  }

  getHealthColor(status: string): string {
    switch (status) {
      case 'healthy': return 'text-green-600';
      case 'warning': return 'text-yellow-600';
      case 'critical': return 'text-red-600';
      default: return 'text-gray-600';
    }
  }

  getHealthBgColor(status: string): string {
    switch (status) {
      case 'healthy': return 'bg-green-50 border-green-200';
      case 'warning': return 'bg-yellow-50 border-yellow-200';
      case 'critical': return 'bg-red-50 border-red-200';
      default: return 'bg-gray-50 border-gray-200';
    }
  }
}

const analyticsService = new AnalyticsService();
export default analyticsService;